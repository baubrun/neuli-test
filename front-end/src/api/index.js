import $ from "jquery";

export const url = "https://reqres.in/api/users";

/**
 * GET users
 * @param {string} v1
 * @param {string} v2
 * @returns Promise
 */
export const fetchUsers = async (v1, v2) => {
  const result = await Promise.all([
    await $.get(`${url}/${v1}`),
    await $.get(`${url}/${v2}`),
  ]);
  return result;
};
