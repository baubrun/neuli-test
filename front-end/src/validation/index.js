import * as yup from "yup";

/**
 * Validate form inputs
 * @param {{var1: string, var2: string}} v
 */
export const validateSchema = (v) => {
  try {
    const schema = yup.object().shape({
      var1: yup.number().required().min(1).max(10),
      var2: yup.number().required().min(1).max(10),
    });
    return schema.validateSync(v);
  } catch (error) {
    return { err: error.message };
  }
};
