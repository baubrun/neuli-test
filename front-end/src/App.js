import React, { Component } from "react";
import "bootstrap/dist/css/bootstrap.min.css";
import Button from "react-bootstrap/Button";
import Form from "react-bootstrap/Form";
import Spinner from "react-bootstrap/Spinner";
import "./App.css";
import { fetchUsers } from "./api";
import Toaster from "./Toaster/Toaster";
import { validateSchema } from "./validation";

class App extends Component {
  constructor(props) {
    super(props);
    // This binding is necessary to make `this` work in the callback
    this.getUsers = this.getUsers.bind(this);
    this.closeMessage = this.closeMessage.bind(this);
    this.handleChange = this.handleChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
    this.isValidForm = this.isValidForm.bind(this);
    this.state = {
      message: "",
      var1: "",
      var2: "",
      user1: "",
      user2: "",
      loading: false,
    };
  }

  closeMessage() {
    this.setState((prev) => ({ ...prev, message: "" }));
  }

  async getUsers() {
    try {
      const [first, second] = await fetchUsers(
        this.state.var1,
        this.state.var2
      );
      this.setState((prev) => ({
        ...prev,
        user1: first.data.first_name,
        user2: second.data.first_name,
      }));
    } catch (error) {
      this.setState((prev) => ({ ...prev, message: error }));
    }
  }

  /**
   * @param {EventTarget} evt
   */
  handleChange(evt) {
    const { name, value } = evt.target;
    this.setState((prev) => ({ ...prev, [name]: value }));
  }

  /**
   * @param {EventTarget} evt
   */
  handleSubmit(evt) {
    evt.preventDefault();
    const validErr = this.isValidForm();
    if (validErr.err) {
      this.setState((prev) => ({ ...prev, message: validErr.err }));
      return;
    }
    this.setState((prev) => ({ ...prev, loading: true }));
    try {
      this.getUsers();
    } catch (error) {
      this.setState((prev) => ({ ...prev, message: error.message }));
    } finally {
      this.setState((prev) => ({ ...prev, loading: false }));
    }
  }

  /**
   * Check is form valid
   * @return {{err: string}}
   */
  isValidForm() {
    const validMessage = validateSchema({
      var1: this.state.var1,
      var2: this.state.var2,
    });
    return validMessage;
  }

  render() {
    return (
      <div className="App">
        <Toaster
          onClose={this.closeMessage}
          show={this.state.message}
          text={this.state.message}
          variant="danger"
        />
        <header className="Simple page">
          <Form onSubmit={this.handleSubmit}>
            <Form.Group controlId="formBasicUser1">
              <Form.Label>User # 1</Form.Label>
              <Form.Control
                type="number"
                placeholder="x"
                name="var1"
                value={this.state.var1}
                onChange={(evt) => this.handleChange(evt)}
                min="1"
                max="10"
              />
            </Form.Group>

            <Form.Group controlId="formBasicUser2">
              <Form.Label>User # 2</Form.Label>
              <Form.Control
                type="number"
                placeholder="y"
                name="var2"
                value={this.state.var2}
                onChange={(evt) => this.handleChange(evt)}
                min="1"
                max="10"
              />
            </Form.Group>
            <Form.Group controlId="formBasicCheckbox">
              <Form.Check type="checkbox" label="Check me out" />
            </Form.Group>
            <Button type="submit">Get Users info</Button>
          </Form>

          <Form>
            {this.state.loading ? (
              <Spinner
                animation="border"
                style={{ marginTop: 5 }}
                variant="primary"
              />
            ) : (
              <Form.Group controlId="formAnswer">
                <Form.Text ref="answer">{`${this.state.user1} ${this.state.user2}`}</Form.Text>
              </Form.Group>
            )}
          </Form>
        </header>
      </div>
    );
  }
}
export default App;
