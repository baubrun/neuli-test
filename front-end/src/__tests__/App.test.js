import React from "react";
import { render } from "@testing-library/react";
import App from "../App";
import { rest } from "msw";
import { setupServer } from "msw/node";
import { url, fetchUsers } from "../api";
import { validateSchema } from "../validation";

const user1 = { data: { first_name: "George" } };
const user2 = { data: { first_name: "Janet" } };

const server = setupServer(
  rest.get(`${url}/1`, (req, res, ctx) => {
    return res(ctx.json(user1));
  }),
  rest.get(`${url}/2`, (req, res, ctx) => {
    return res(ctx.json(user2));
  })
);

describe("App", () => {
  it("should match snapshot", () => {
    const { container: containerOne } = render(<App />);
    expect(containerOne).toMatchSnapshot();
  });
});

describe("GET requests", () => {
  beforeAll(() => {
    server.listen();
  });

  afterEach(() => {
    server.resetHandlers();
  });

  afterAll(() => {
    server.close();
  });

  it("GET requests should return first names ", async () => {
    const [first, second] = await fetchUsers("1", "2");
    expect([first, second]).toEqual([user1, user2]);
  });
});

describe("validateSchema", () => {
  it("should throw error if params falsy ", async () => {
    try {
      validateSchema({ var1: "", var2: undefined });
    } catch (error) {
      const e = "var2 is a required field";
      expect(error.message).toEqual(e);
    }
  });

  it("should throw error if  param < 1 or param > 10  ", async () => {
    try {
      validateSchema({ var1: "11", var2: "2" });
    } catch (error) {
      const e = "var1 must be less than or equal to 10";
      expect(error.message).toEqual(e);
    }
  });
});
