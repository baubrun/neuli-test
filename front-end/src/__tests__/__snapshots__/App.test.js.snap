// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match snapshot 1`] = `
<div>
  <div
    class="App"
  >
    <header
      class="Simple page"
    >
      <form
        class=""
      >
        <div
          class="form-group"
        >
          <label
            class="form-label"
            for="formBasicUser1"
          >
            User # 1
          </label>
          <input
            class="form-control"
            id="formBasicUser1"
            max="10"
            min="1"
            name="var1"
            placeholder="x"
            type="number"
            value=""
          />
        </div>
        <div
          class="form-group"
        >
          <label
            class="form-label"
            for="formBasicUser2"
          >
            User # 2
          </label>
          <input
            class="form-control"
            id="formBasicUser2"
            max="10"
            min="1"
            name="var2"
            placeholder="y"
            type="number"
            value=""
          />
        </div>
        <div
          class="form-group"
        >
          <div
            class="form-check"
          >
            <input
              class="form-check-input"
              id="formBasicCheckbox"
              type="checkbox"
            />
            <label
              class="form-check-label"
              for="formBasicCheckbox"
              title=""
            >
              Check me out
            </label>
          </div>
        </div>
        <button
          class="btn btn-primary"
          type="submit"
        >
          Get Users info
        </button>
      </form>
      <form
        class=""
      >
        <div
          class="form-group"
        >
          <small
            class="form-text"
          >
             
          </small>
        </div>
      </form>
    </header>
  </div>
</div>
`;
